CMakeLists.txt

```diff
@@ -0,0 +1,345 @@
+		set(TOOL_NAME sound-export)
+		set(FOLDER_NAME sounds/)
+		set(OUT1 .s)
+		set(OUT2 .hpp)
+	endif()
+	set(RESOURCE_FILES ${ARGV2})
+	set(IN2 ${ARGV3})
+
+	#make a directory in build folder for the output files
+	file(MAKE_DIRECTORY ${CMAKE_BINARY_DIR}/data/${FOLDER_NAME})
+
+	set(OBJECT_LIST_NAME ${PLURAL}_OBJECT_LIST)
+	set(OUT1_LIST_NAME ${PLURAL}_OUT1_LIST)
+	set(OUT2_LIST_NAME ${PLURAL}_OUT2_LIST)
+
+	foreach(FILE ${RESOURCE_FILES})
+		get_filename_component(file_dir "${FILE}" PATH)
+		string(REPLACE "." ";." file ${FILE})
+		list(GET file 0 file_name)
+		list(GET file 1 file_type)
+
+		set(OUTPUT_FILE_1 ${CMAKE_BINARY_DIR}/data/${FOLDER_NAME}${file_name}${OUT1})
+		set(OUTPUT_FILE_2 ${CMAKE_BINARY_DIR}/data/${FOLDER_NAME}${file_name}${OUT2})
+
+		# This command generates the files from the source/data files
+		# OUTPUT means the files are set as Generated in cmake,
+		# so when they are used as a source, a dependency is set on their generation
+		# We pass in the dependencies to DEPENDS, the first call of this function passes in the Tools project as a dependency,
+		# so that all subsequent commands and targets are implicitely dependent on Tools having finished
+		# WORKING_DIRECTORY is set to the project root (${CMAKE_SOURCE_DIR}) so a relative path to the file that is to be exported
+		# is sent to the tool, for proper generation of namespaces
+		add_custom_command(OUTPUT ${OUTPUT_FILE_1} ${OUTPUT_FILE_2}
+			DEPENDS ${TARGET_DEPENDENCIES}
+			COMMAND ${CMAKE_BINARY_DIR}/tools/tools ${TOOL_NAME}
+				./data/${FOLDER_NAME}${file_name}${file_type}
+				${OUTPUT_FILE_1} ${OUTPUT_FILE_2} ${IN2}
+			WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+			COMMENT "Exporting ${RESOURCE_NAME} ${file_name}")
+
+		# We don't want to generate and object and dependency? file for audio.json, since it only outputs a header and another json file
+		if(NOT ${RESOURCE_NAME} MATCHES "SETTING")
+			file(MAKE_DIRECTORY ${CMAKE_BINARY_DIR}/data/${FOLDER_NAME}/${file_name})
+
+			set(OUTPUT_OBJECT ${CMAKE_BINARY_DIR}/data/${FOLDER_NAME}${file_name}/${file_name}.o)
+			set(OUTPUT_OBJECT_D ${CMAKE_BINARY_DIR}/data/${FOLDER_NAME}${file_name}/${file_name}.d)
+
+			# This command generates the object file and dependency? of the exported ASM and header
+			# OUTPUT again means these files are set as GENERATED
+			# MAIN_DEPENDENCY sets the input ASM file as the main dependency
+			# DEPENDS sets the header file as a dependency
+			# We need too set dependencies to the input files used in the command, as it is not automatically done even though they are source files.
+			# This ensures this command is ran after generating the ASM and header files
+			add_custom_command(OUTPUT ${OUTPUT_OBJECT} ${OUTPUT_OBJECT_D}
+				MAIN_DEPENDENCY ${OUTPUT_FILE_1}
+				DEPENDS ${OUTPUT_FILE_2}
+				COMMAND ${GBA_C_COMPILER} -MMD -MP -MF ${OUTPUT_OBJECT_D} -x assembler-with-cpp ${ASFLAGS} -c ${OUTPUT_FILE_1} -o ${OUTPUT_OBJECT}
+				COMMENT "BUILDING ${OUTPUT_OBJECT} AND ${OUTPUT_OBJECT_D}")
```

Comment:
Here, I wouldn't know how to split the list between file 1 and file 2, but adding them to the final build list should suffice? (I guess you might already have experimented with that).